### PR TITLE
feat(mito): Flush region

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR implements most of flush related methods of the worker and supports flushing a region manually.
- Triggers flush if the memory usage of the engine is larger than `global_write_buffer_size`. It picks some regions that are not flushed after `auto_flush_interval` and a region with the largest memtable to flush.
- Rejects write if the memory usage is larger than `global_write_buffer_reject_size`.
- Each region uses `AccessLayer` to read/write/delete SSTs.
- `FlushScheduler` manages flush requests of regions and schedules flush tasks.
- Tracking memory usage will be implemented in #2266
- Stalling or delaying write requests is not supported yet. I plan to implement it later.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #1869 
- #2296
- #2266

### CODE REVIEW

File: src/mito2/src/manifest/action.rs [hunk 1]
```diff
@@ -217,7 +226,7 @@ mod tests {
     // modification to manifest-related structs is compatible with older manifests.
     #[test]
     fn test_region_manifest_compatibility() {
-        let region_edit = r#"{"region_version":0,"flushed_sequence":null,"files_to_add":[{"region_id":4402341478400,"file_name":"4b220a70-2b03-4641-9687-b65d94641208.parquet","time_range":[{"value":1451609210000,"unit":"Millisecond"},{"value":1451609520000,"unit":"Millisecond"}],"level":1}],"files_to_remove":[{"region_id":4402341478400,"file_name":"34b6ebb9-b8a5-4a4b-b744-56f67defad02.parquet","time_range":[{"value":1451609210000,"unit":"Millisecond"},{"value":1451609520000,"unit":"Millisecond"}],"level":0}]}"#;
+        let region_edit = r#"{"region_version":0,"flushed_entry_id":null,"files_to_add":[{"region_id":4402341478400,"file_name":"4b220a70-2b03-4641-9687-b65d94641208.parquet","time_range":[{"value":1451609210000,"unit":"Millisecond"},{"value":1451609520000,"unit":"Millisecond"}],"level":1}],"files_to_remove":[{"region_id":4402341478400,"file_name":"34b6ebb9-b8a5-4a4b-b744-56f67defad02.parquet","time_range":[{"value":1451609210000,"unit":"Millisecond"},{"value":1451609520000,"unit":"Millisecond"}],"level":0}]}"#;
```
Comment (waynexia) [inline src/mito2/src/manifest/action.rs:229]: I think we should update this text

File: src/mito2/src/region.rs [hunk 2]
```diff
@@ -82,6 +83,17 @@ impl MitoRegion {
         let version_data = self.version_control.current();
         version_data.version
     }
+
+    /// Returns last flush timestamp in millis.
+    pub(crate) fn last_flush_millis(&self) -> i64 {
```
Comment (waynexia) [inline src/mito2/src/region.rs:88]:
```suggestion
    pub(crate) fn last_flush_millis(&self) -> Millisecond {
```

File: src/mito2/src/worker.rs [hunk 3]
```diff
@@ -398,12 +401,15 @@ impl<S: LogStore> RegionWorkerLoop<S> {
...
                 DdlRequest::Open(req) => self.handle_open_request(ddl.region_id, req).await,
                 DdlRequest::Close(_) => self.handle_close_request(ddl.region_id).await,
-                DdlRequest::Alter(_)
-                | DdlRequest::Drop(_)
-                | DdlRequest::Flush(_)
-                | DdlRequest::Compact(_) => todo!(),
+                DdlRequest::Alter(_) => todo!(),
+                DdlRequest::Flush(_) => {
+                    self.handle_flush_request(ddl.region_id, ddl.sender).await;
+                    continue;
```
Comment (waynexia) [inline src/mito2/src/worker.rs:410]: what's this for?

File: src/mito2/src/sst/version.rs [hunk 4]
```diff
@@ -40,13 +41,45 @@ impl SstVersion {
...
+        &mut self,
+        file_purger: FilePurgerRef,
+        files_to_add: impl Iterator<Item = FileMeta>,
+    ) {
+        for file in files_to_add {
+            let level = file.level;
+            let handle = FileHandle::new(file, file_purger.clone());
+            self.levels[level as usize]
+                .files
+                .insert(handle.file_id(), handle);
```
Comment (waynexia) [inline src/mito2/src/sst/version.rs:59]: Should we check if this file is already exist?

File: src/mito2/src/worker.rs (hunk 3 above)
Comment (evenyag) [reply src/mito2/src/worker.rs:410]: It moves the `ddl.sender`. I can't use the sender after the match expression again so I have to continue here.

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/mito2/src/sst/parquet/writer.rs [hunk 5]
```diff
@@ -125,4 +127,31 @@ impl ParquetWriter {
...
 }
 
+#[derive(Default)]
+struct SourceStats {
+    /// Number of rows fetched.
+    num_rows: usize,
+    /// Min timestamp from fetched batches.
+    min_timestamp: Option<Timestamp>,
+    /// Max timestamp from fetched batches.
+    max_timestamp: Option<Timestamp>,
```
Comment (v0y4g3r) [inline src/mito2/src/sst/parquet/writer.rs:137]: Maybe merge `min_timestamp` and `max_timestamp` into `time_range: Option<(Timestamp, Timestamp)>` as these two field can only be either all present or not.

File: src/mito2/src/worker/handle_flush.rs [hunk 6]
```diff
@@ -14,81 +14,167 @@
...
+
+            if region.last_flush_millis() < min_last_flush_time {
+                // If flush time of this region is earlier than `min_last_flush_time`, we can flush this region.
+                let task = self.new_flush_task(region, FlushReason::EngineFull);
+                self.flush_scheduler.schedule_flush(region, task)?;
+            }
+        }
+
+        // Flush memtable with max mutable memtable.
+        if let Some(region) = max_mem_region {
```
Comment (v0y4g3r) [inline src/mito2/src/worker/handle_flush.rs:158]: Looks like this method only flushes the region with max memtable? I wonder to what extent does this method help to reduce write buffer usage.

File: src/mito2/src/flush.rs [hunk 7]
```diff
@@ -90,10 +86,10 @@ impl WriteBufferManager for WriteBufferManagerImpl {
 pub enum FlushReason {
     /// Other reasons.
     Others,
-    /// Memtable is full.
-    MemtableFull,
     /// Engine reaches flush threshold.
     EngineFull,
+    /// Manual flush.
+    Manual,
     // TODO(yingwen): Alter, manually.
```
Comment (v0y4g3r) [inline src/mito2/src/flush.rs:93]: `Manual`'s already added.

**Review by v0y4g3r — APPROVED** [review]
Mostly LGTM

File: src/mito2/src/worker/handle_flush.rs (hunk 6 above)
Comment (evenyag) [reply src/mito2/src/worker/handle_flush.rs:158]: We can optimize this later.

**Pull request merged**